ttnn/cpp/ttnn/tensor/tensor_utils.cpp

```diff
@@ -210,6 +210,229 @@ Tensor convert_conv_weight_tensor_to_tiled_layout(
+                    }
+                }
+            }
+        }
+        if constexpr (std::is_same<T, float>::value) {
+            if (output_dtype == DataType::BFLOAT8_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp8_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
+            if (output_dtype == DataType::BFLOAT4_B) {
+                auto output_float_data = output_buffer.get();
+                auto output_packed_data =
+                    pack_fp32_vec_as_bfp4_tiles(output_float_data, /*row_major_input=*/false, /*is_exp_a=*/false);
+                auto output_uint32_buffer = owned_buffer::create<uint32_t>(std::move(output_packed_data));
+                auto rm_tensor = Tensor(
+                    std::move(OwnedStorage{std::move(output_uint32_buffer)}),
+                    output_shape,
+                    output_dtype,
+                    Layout::ROW_MAJOR);
+                return rm_tensor.to(Layout::TILE);
+            }
+        } else {
+            TT_ASSERT((output_dtype != DataType::BFLOAT8_B) || (output_dtype != DataType::BFLOAT4_B));
+        }
+        auto rm_tensor =
+            Tensor(std::move(OwnedStorage{std::move(output_buffer)}), output_shape, output_dtype, Layout::ROW_MAJOR);
+        return rm_tensor.to(Layout::TILE);
+    };
+
+    auto convert_tensor = [&compute](const auto& conv_weight_tensor) {
+        return std::visit(
+            [&compute](auto&& storage) -> Tensor {
+                using StorageType = std::decay_t<decltype(storage)>;
+                if constexpr (std::is_same_v<StorageType, OwnedStorage>) {
+                    return compute(owned_buffer::get_as<T>(storage.buffer));
+                } else if constexpr (std::is_same_v<StorageType, BorrowedStorage>) {
+                    return compute(borrowed_buffer::get_as<T>(storage.buffer));
+                } else {
+                    TT_THROW("Unsupported storage type");
+                }
+            },
+            conv_weight_tensor.get_storage());
+    };
+    return ttnn::distributed::is_multi_device_tensor(conv_weight_tensor) ? transform(conv_weight_tensor, convert_tensor)
+                                                                         : convert_tensor(conv_weight_tensor);
+}
+
+
+// Converts convolution weights to tilized 2d matrix layout for block sharded conv.
+// Returns a new tensor with layout=Tile
+Tensor convert_conv_weight_tensor_to_tiled_layout_block_sharded(
```

Comment:
Any chance to do it on device?
If not please create c++ test for this function.

Reply:
As of now we are doing it on host. 
There are not tests for any function in this file(they are tested indirectly by conv2d op). I shall create separate issue for same.

Reply:
There is a separate tracking for moving all weight pre-processing to device: #14080

Reply:
Added test cases, pls check.